persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java

```diff
@@ -0,0 +1,131 @@
+import org.infinispan.persistence.remote.configuration.global.RemoteContainersConfiguration;
+import org.infinispan.persistence.remote.global.GlobalRemoteContainers;
+import org.infinispan.persistence.remote.logging.Log;
+import org.infinispan.util.concurrent.BlockingManager;
+
+/**
+ * @since 15.0
+ **/
+@Scope(Scopes.GLOBAL)
+public class GlobalRemoteContainersImpl implements GlobalRemoteContainers {
+
+   @Inject
+   GlobalConfiguration globalConfiguration;
+
+   @Inject
+   BlockingManager blockingManager;
+
+   private final Map<String, CompletionStage<RemoteCacheManager>> remoteContainers = new ConcurrentHashMap<>();
+
+   @Start
+   public void start() {
+      RemoteContainersConfiguration configuration = globalConfiguration.module(RemoteContainersConfiguration.class);
+      if (configuration != null) {
+         for (Map.Entry<String, RemoteContainerConfiguration> e : configuration.configurations().entrySet()) {
+            remoteContainers.put(e.getKey(), CompletableFutures.completedNull());
+         }
+      }
+   }
+
+   @Stop
+   public void stop() {
+      remoteContainers.values().forEach(c -> c.thenApply(this::stop));
+   }
+
+   private Void stop(RemoteCacheManager rcm) {
+      if (rcm != null) rcm.stop();
+      return null;
+   }
+
+   private void remoteCacheManagerShutdown(String name) {
+      remoteContainers.computeIfPresent(name, (key, prev) -> prev.thenApply(ignore -> null));
+   }
+
+   @Override
+   public CompletionStage<RemoteCacheManager> cacheContainer(String name, Marshaller marshaller) {
+      CompletionStage<RemoteCacheManager> cs = remoteContainers.computeIfPresent(name, (k, cf) ->
+            cf.thenCompose(prev -> prev != null
+                  ? CompletableFuture.completedFuture(((RefCountedRemoteCacheManager) prev).increment())
+                  : createCacheManager(k, marshaller)));
+
+      if (cs == null) {
+         return CompletableFuture.failedFuture(Log.CONFIG.unknownRemoteCacheManagerContainer(name));
+      }
+
+      return cs.thenApply(rcm -> {
+         Marshaller inUse = rcm.getConfiguration().marshaller();
+         if (!inUse.equals(marshaller))
+            throw Log.CONFIG.shouldUseSameMarshallerWithContainer(inUse.getClass(), marshaller.getClass());
+         return rcm;
+      });
+   }
+
+   private CompletionStage<RemoteCacheManager> createCacheManager(String name, Marshaller marshaller) {
+      RemoteContainersConfiguration configuration = globalConfiguration.module(RemoteContainersConfiguration.class);
+      if (configuration != null) {
+         RemoteContainerConfiguration rcc = configuration.configurations().get(name);
+
+         if (rcc == null)
+            return CompletableFuture.failedFuture(new IllegalStateException("No configuration defined for container " + name));
+
+         ConfigurationBuilder builder = new ConfigurationBuilder();
+         builder.withProperties(rcc.properties());
```

Comment:
We should make sure to add some tests to make sure these allow same configuration options as the cache store, especially security based ones.

Reply:
Updated.